// unframer/dist/cli.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cli = void 0;
const events_1 = require("events");
const json5_1 = __importDefault(require("json5"));
const exporter_js_1 = require("./exporter.js");
const api_client_js_1 = require("./generated/api-client.js");
const cac_1 = require("cac");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importStar(require("path"));
const utils_js_1 = require("./utils.js");
const configNames = ['unframer.config.json', 'unframer.json'];
exports.cli = (0, cac_1.cac)('unframer');
let defaultOutDir = 'framer';
exports.cli.command('[projectId]', 'Run unframer with optional project ID')
    .option('--outDir <dir>', 'Output directory', { default: defaultOutDir })
    .option('--external [package]', 'Make some package external, do not pass a package name to make all packages external', {
    default: false,
})
    .option('--watch', 'Watch for changes and rebuild', { default: false })
    .option('--debug', 'Enable debug logging', { default: false })
    .action(async function main(projectId, options) {
    const external_ = options.external;
    const allExternal = external_ === true;
    const externalPackages = Array.isArray(external_)
        ? external_.filter((x) => x.trim())
        : typeof external_ === 'string'
            ? [external_]
            : [];
    try {
        if (options.debug) {
            utils_js_1.logger.debug = true;
        }
        const outDir = options.outDir;
        const controller = new AbortController();
        const signal = controller.signal;
        const watch = options.watch;
        if (projectId) {
            utils_js_1.logger.log(`Fetching config for project ${projectId}`);
            const client = await (0, api_client_js_1.createClient)({
                url: process.env.UNFRAMER_SERVER_URL ||
                    'https://unframer.co',
            });
            const { data, error } = await client.api.plugins.reactExportPlugin
                .project({ projectId })
                .get();
            if (error) {
                utils_js_1.logger.error('Error fetching project data:', error);
                throw error;
            }
            const websiteUrl = data?.project?.websiteUrl;
            utils_js_1.logger.log('unframer data', data);
            const projectName = data?.project?.projectName || '';
            if (projectName) {
                utils_js_1.spinner.info(`Using project: ${projectName}`);
            }
            let cwd = path_1.default.resolve(process.cwd(), outDir || 'framer');
            utils_js_1.logger.log('bundling', cwd);
            const { rebuild, buildContext } = await (0, exporter_js_1.bundle)({
                config: {
                    outDir,
                    externalPackages,
                    allExternal,
                    projectId: data?.project?.projectId,
                    projectName,
                    fullFramerProjectId: data?.project?.fullFramerProjectId,
                    locales: data?.locales,
                    components: Object.fromEntries(await Promise.all(data.components.map(async (c) => [
                        (0, utils_js_1.componentNameToPath)(c.name),
                        c.url,
                    ]))),
                    componentBreakpoints: data.breakpoints
                        ?.map((b) => {
                        const c = data.components.find((c) => c.id === b.componentId);
                        if (!c) {
                            return;
                        }
                        return {
                            ...b,
                            componentName: (0, utils_js_1.componentNameToPath)(c.name),
                        };
                    })
                        .filter(utils_js_1.isTruthy) || [],
                    tokens: data.colorStyles,
                    framerWebPages: data.framerWebPages || [],
                },
                watch,
                cwd,
                signal,
            });
            // console.log('buildContext', buildContext)
            if (!websiteUrl || !options.watch) {
                await buildContext?.dispose?.();
                return;
            }
            utils_js_1.spinner.start(`Waiting for changes, try editing a Framer component and publishing...`);
            let lastEtag = null;
            const startTime = Date.now();
            while (Date.now() - startTime < 30 * 60 * 1000) {
                const etag = await fetch(websiteUrl, {
                    method: 'HEAD',
                })
                    .then((response) => response.headers.get('etag'))
                    .catch((error) => {
                    utils_js_1.logger.error('Error fetching etag:', error);
                    return null;
                });
                utils_js_1.logger.log('etag', etag);
                if (etag && lastEtag && etag !== lastEtag) {
                    utils_js_1.spinner.start(`Detected Framer website change, rebuilding...`);
                    lastEtag = etag;
                    await rebuild();
                }
                if (etag) {
                    lastEtag = etag;
                }
                await (0, utils_js_1.sleep)(1000 * 2);
            }
        }
        // legacy behavior without Framer plugin
        fixOldUnframerPath();
        const cwd = process.cwd();
        utils_js_1.logger.log(`Looking for ${configNames.join(', ')} in ${cwd}`);
        const { findUp } = await import('find-up');
        const configPath = await findUp(configNames, { cwd });
        if (!configPath) {
            utils_js_1.logger.log(`No ${configNames.join(', ')} found`);
            return;
        }
        const configBasename = (0, path_1.basename)(configPath);
        const configContent = fs_1.default.readFileSync(configPath, 'utf8');
        if (!configContent) {
            utils_js_1.logger.log(`No ${configBasename} contents found`);
            return;
        }
        const config = json5_1.default.parse(configContent);
        if (outDir !== defaultOutDir) {
            config.outDir = outDir;
        }
        (0, events_1.setMaxListeners)(0, controller.signal);
        const { buildContext } = await (0, exporter_js_1.bundle)({
            config: { ...config, externalPackages, allExternal },
            watch,
            signal: controller.signal,
            cwd: path_1.default.resolve(process.cwd(), config.outDir || 'framer'),
        });
        await buildContext.dispose?.();
    }
    catch (error) {
        utils_js_1.logger.error('Error in main:', error);
        throw error;
    }
});
const defaultConfig = `{
    "schema": "https://unframer-schema.vercel.app/schema.json",
    "outDir": "./framer",
    "components": {
        // add here your Framer components urls, the code will be written to outDir/{componentName}.js
        "example-hero": "https://framer.com/m/Header-WtSW.js",
    }
}
`;
function fixOldUnframerPath() {
    // if unframer.json exists, rename it to unframer.config.json
    const oldConfigPath = fs_1.default.existsSync('unframer.json');
    if (oldConfigPath) {
        fs_1.default.renameSync('unframer.json', 'unframer.config.json');
        utils_js_1.logger.green('legacy unframer.json config renamed to unframer.config.json');
        return true;
    }
    return false;
}
const version = require('../package.json').version;
exports.cli.version(version).help();
exports.cli.command('init', 'Init the unframer.config.json config').action(async (options) => {
    let fixed = fixOldUnframerPath();
    if (fixed) {
        return;
    }
    fs_1.default.writeFileSync(`unframer.config.json`, defaultConfig);
    const p = path_1.default.resolve(process.cwd(), 'unframer.config.json');
    console.log(`${p} file created`);
});
function safeJsonParse(json) {
    try {
        return json5_1.default.parse(json);
    }
    catch (e) {
        return null;
    }
}
function pluck(o, names) {
    return Object.fromEntries(names.map((n) => [n, o[n]]));
}
function getNewNames(oldConfig, newConfig) {
    // get the new names, also check if the previous url (object value) has changed
    const oldKeys = Object.keys(oldConfig.components);
    const newKeys = Object.keys(newConfig.components);
    const newNames = newKeys.filter((key) => {
        if (!oldKeys.includes(key)) {
            return true;
        }
        if (oldConfig.components[key] !== newConfig.components[key]) {
            return true;
        }
        return false;
    });
    return newNames;
}
